x-pack/plugins/encrypted_saved_objects/server/lib/encrypted_saved_objects_service.ts

```diff
@@ -0,0 +1,218 @@
+   * @param id Id of the saved object to strip encrypted attributes from.
+   * @throws Will throw if decryption fails for whatever reason.
+   * @throws Will throw if any of the attributes to decrypt is not a string.
+   */
+  public async decryptAttributes<T extends Record<string, unknown>>(
+    type: string,
+    attributes: T,
+    id: string
+  ) {
+    const typeRegistration = this.typeRegistrations.get(type);
+    if (typeRegistration === undefined) {
+      return;
+    }
+
+    const encryptionAAD = this.getAAD(typeRegistration, attributes);
+    const notDefinedAttributesToDecrypt = [];
+    for (const attributeName of typeRegistration.attributesToEncrypt) {
+      const attributeValue = attributes[attributeName];
+      if (attributeValue == null) {
+        // Normally we expect all encrypted attributes to be defined and have a `string` type, but
+        // if it's not the case we should collect and log them to make troubleshooting easier.
+        notDefinedAttributesToDecrypt.push(attributeName);
+      } else if (typeof attributeValue !== 'string') {
+        this.logger.audit.decryptAttributeFailure(attributeName, type, id);
+        throw new Error(
+          `Encrypted "${attributeName}" attribute should be a string, but found ${typeDetect(
+            attributeValue
+          )}`
+        );
+      } else {
+        try {
+          attributes[attributeName] = await this.crypto.decrypt(attributeValue, encryptionAAD);
+        } catch (err) {
+          this.logger.audit.decryptAttributeFailure(attributeName, type, id);
+          throw err;
+        }
+      }
+    }
+
+    if (notDefinedAttributesToDecrypt.length > 0) {
+      this.logger.debug(
+        `The following attributes of saved object "${type}:${id}" were expected to be decrypted, but were not defined: ${notDefinedAttributesToDecrypt}`
+      );
+    }
+
+    this.logger.audit.decryptAttributesSuccess(typeRegistration.attributesToEncrypt, type, id);
+  }
+
+  /**
+   * Generates string representation of the Additional Authenticated Data based on the specified saved
+   * object type and attributes.
+   * @param typeRegistration Saved object type registration parameters.
+   * @param attributes All attributes of the saved object instance of the specified type.
+   */
+  private getAAD<T extends Record<string, unknown>>(
+    typeRegistration: EncryptedSavedObjectTypeRegistration,
+    attributes: T
+  ) {
+    // Collect all attributes (both keys and values) that should contribute to AAD.
+    const attributesAAD = Array.from(Object.entries(attributes)).filter(([attributeKey]) => {
+      return (
+        !typeRegistration.attributesToEncrypt.has(attributeKey) &&
+        (typeRegistration.attributesToExcludeFromAAD == null ||
+          !typeRegistration.attributesToExcludeFromAAD.has(attributeKey))
+      );
+    });
+
+    // TODO: We can't include `id` into AAD since it may not be specified when saved object is created.
```

Comment:
__note:__ unless we force consumers to specify `id` when they create saved objects with encrypted fields.

Reply:
I have a WIP commit where we generate id if it's not provided inside of saved object client wrapper, but would like to hear your thoughts first.

Reply:
During one of the earlier proposals, I suggested using a uuid v4 backed by a cryptographically secure random number, to mirror what the Elasticsearch team does when generating document IDs for things like tokens. This was to provide some additional security from malicious actors being able to guess IDs. While we shouldn't be relying on the sole protection of a user being unable to guess an ID, it could potentially help the security in some situations. If we were to do so here, then we would need to generate our own IDs for create operations, and then we could use it as part of the AAD.

Reply:
> During one of the earlier proposals, I suggested using a uuid v4 backed by a cryptographically secure random number, to mirror what the Elasticsearch team does when generating document IDs for things like tokens. This was to provide some additional security from malicious actors being able to guess IDs.

I thought it was related to a dedicated "secret" saved objects, but now we're leaning towards encrypted attributes on the existing/known saved objects where `id` is allowed to be specified by the consumer (e.g. through `options` argument for public `SavedObjectClient.create` method). That's why my initial thought was to generate `id` only if it's not specified by the consumer and leverage the same method from SavedObjectSerializer that generate ids so that they follow the same convention (currently relies on `uuid.v1()`). Let's sync offline on that.